### ISSUE TITLE
Intraprocess multithreaded test is flaky

### ISSUE THREAD

**Issue by gerkey** [ros2/system_tests#92]
(Not sure whether this is the same problem as #39.)

Below is the console output from a failure seen in a nightly build on OSX, running the multithreaded test with connext.

Looks like this is the sequence of events (https://github.com/ros2/system_tests/blob/master/test_rclcpp/test/test_multithreaded.cpp#L78-L110):

* Publish one message, spin, and expect it to have been received by all subscribers. Instead, nobody has yet received it (failure on line 84).
* Spin more and expect that all subscribers will still have received only that first message. Instead, by this time, 1 subscriber has received it (failure on line 88).
* Reset the counter, publish 5 more messages, and expect that each subscriber will have received exactly 5 messages. Instead, both those 5 new messages and the 3 pending message from earlier are received (failure on line 110).

I'm not sure of the correct fix here. Increase the heuristic delay at the beginning (https://github.com/ros2/system_tests/blob/master/test_rclcpp/test/test_multithreaded.cpp#L72)? Spin more before checking the initial delivery (https://github.com/ros2/system_tests/blob/master/test_rclcpp/test/test_multithreaded.cpp#L83)? Something else?

~~~
[ RUN      ] test_multithreaded__.multi_consumer_intra_process
RTI Data Distribution Service Evaluation License issued to OSRF (OSRF01) dthomas@osrfoundation.org For non-production use only.
Expires on 08-jun-2016 See www.rti.com for more information.
/Users/osrf/jenkins/workspace/nightly_osx/ws/src/ros2/system_tests/test_rclcpp/test/test_multithreaded.cpp:84: Failure
Value of: subscriptions.size()
  Actual: 4
Expected: counter.load()
Which is: 0
callback()    1 with message data 1
/Users/osrf/jenkins/workspace/nightly_osx/ws/src/ros2/system_tests/test_rclcpp/test/test_multithreaded.cpp:88: Failure
Value of: subscriptions.size()
  Actual: 4
Expected: counter.load()
Which is: 1
callback()    1 with message data 1
callback()    2 with message data 1
callback()    3 with message data 1
callback()    4 with message data 1
callback()    5 with message data 1
callback()    6 with message data 1
callback()    7 with message data 1
callback()    8 with message data 2
callback()    9 with message data 2
callback()   10 with message data 2
callback()   11 with message data 2
callback()   12 with message data 3
callback()   13 with message data 3
callback()   14 with message data 3
callback()   15 with message data 3
callback()   16 with message data 4
callback()   17 with message data 4
callback()   18 with message data 4
callback()   19 with message data 4
callback()   20 with message data 5
callback()   21 with message data 5
callback()   22 with message data 5
callback()   23 with message data 5
/Users/osrf/jenkins/workspace/nightly_osx/ws/src/ros2/system_tests/test_rclcpp/test/test_multithreaded.cpp:110: Failure
Value of: 5 * subscriptions.size()
  Actual: 20
Expected: counter.load()
Which is: 23
~~~

**Comment by gerkey** [comment]
I can reproduce it locally if I run `stress` in parallel, e.g.:

~~~
# sudo apt-get install stress
stress -c 8 &
./build/test_rclcpp/gtest_multithreaded__rmw_connext_cpp
~~~

I'll try some things to fix it.

**Comment by gerkey** [comment]
Adding a 5s sleep after the first publish (https://github.com/ros2/system_tests/blob/master/test_rclcpp/test/test_multithreaded.cpp#L78) fixes the problem (but is obviously a bad idea). I need to get some education about what should be expected from `spin_some()`, particularly in the intraprocess case.

**Comment by gerkey** [comment]
@wjwwood, any thoughts on what is / should be happening here?

**Comment by gerkey** [comment]
Should be fixed by #94.
